### ISSUE TITLE
identifyParticles() | subscript out of bounds

### ISSUE THREAD

**Issue by haasdavid** [marjoleinbruijning/trackdem#13]
Dear Marjolein,

while I could solve most issues, one error still orrours every now and then. I cannot tell where it is coming from. 
The 4th step of `identifyParticles` sometimes outputs the following error. This makes it difficult to analyse >200 datasets in a for loop.

Do you need any dummy-data to better identify the issue?

> partIden <- identifyParticles(sbg=allImages,
>                                threshold = 0.2, # store value in "thres" below
>                                pixelRange=c(5,500), # store in "pxrange" below
>                                autoThres=FALSE,
>                                select = "light")
>  	 Particle Identification: Particle statistics (4 out of 5)           Error in `[<-`(`*tmp*`, loc, 1:5, value = .Call("projectedPS", A[, , i],  : 
>   subscript out of bounds
> 

Kind regards,
David

**Comment by marjoleinbruijning** [comment]
Hi David,

please do send me some dummy data the produces the error (you can email me
at ***@***.***) and I will have a look!

Best, Marjolein

Op wo 26 mei 2021 om 14:41 schreef haasdavid ***@***.***>:

**Comment by haasdavid** [comment]
Dear majorlein,

 

thank you once again for your fast reply.

 

Following this link, you will find the dataset that created the error:

 <https://1drv.ms/u/s!AlJI76ZyXH_Xg6UHnZc8buc5pNnrsw?e=LVrfBr> https://1drv.ms/u/s!AlJI76ZyXH_Xg6UHnZc8buc5pNnrsw?e=LVrfBr

 

I used the frames 100-300. 

 

partIden parameters and error message:

 

partIden <- identifyParticles(sbg=allImages,
threshold = 0.2, # store value in "thres" below
pixelRange=c(5,500), # store in "pxrange" below
autoThres=FALSE,
select = "light")
Particle Identification: Particle statistics (4 out of 5) Error in [<-(*tmp*, loc, 1:5, value = .Call("projectedPS", A[, , i], :
subscript out of bounds

 

Similar datasets I want to batch process seem to work. The error occurs on an irregular basis.

 

Best regards,

David

**Comment by marjoleinbruijning** [comment]
Hi David,

The error arises because in one or more of the frames, there are zero
particles detected. *Trackdem* can currently not handle this (and there is
no easy workaround for this issue, as you will run intro problems with the
tracking). I updated the identifyParticles() function so that works fine,
but it returns a warning, saying that the tracking won't work. The function
trackParticles() now returns an error stating the problem.

In your image sequence, I indeed see many frames without any particles.
Maybe you can manually clip your image sequence, to only include the period
during which the particles are visible?


If you're using a loop to analyze multiple image sequences, you could also
include an if statement after you run identifyParticles(), to test if all
frames contain > 0 particles, before you continue with your tracking
function, for instance using something like: if (all(summary(partIden1)$n >
0)) {...}

Please download the latest version from Github, I haven't submitted it to
CRAN.

Marjolein

Op vr 28 mei 2021 om 12:22 schreef haasdavid ***@***.***>:

**Comment by haasdavid** [comment]
Hi Marjolein,

thank you for your response and the suggested workaround. I will try that!

I will have several hundred image sequences (500-1000 sequences), that I want to batch process.
All sequences will contain particles moving through the AOI following a slope. If I got you right, especially at the end - when all particles have left the field of view or remain still - trackdem will run into the error?

Is there a way to tell trackdem to only include images with particles or stop, when the next x frames have no particles in them? So that it only uses the frames with moving particles that git detected up to this point? I Think that would solve all the issues at the current state. 

Tank you once again and with best regards,
David

Ps.: sorry... I wrote you name wrong in the previous messages. Won't happen again!

**Comment by marjoleinbruijning** [comment]
Hi David,

There are different ways to do this, try for example something like this:

# Option 1: use all images with more than 0 particles
abovezero <- summary(partIden1)$n > 0 # select frames with > 0 particles
allFullImages1 <- loadImages (dirPictures= "xxx"
                             ,nImages = which(abovezero)) # load images
again, but now only that have > 0 particles
allImages1 <- subtractBackground(bg = stillBack1) # use 'old' background
partIden1 <- identifyParticles(sbg = allImages1,threshold = 0.2,
                              select = "light",
                              pixelRange = c(5,500))
summary(partIden1)$n # all larger than 0

# Option 2: use all frames up to first frame with 0 particles
abovezero <- summary(partIden1)$n > 0 # select frames with > 0 particles
untilzero <- min(which(!abovezero))-1 # obtain first frame with 0 particles
allFullImages1 <- loadImages (dirPictures= "xxx"
                             ,nImages = 1:untilzero)
allImages1 <- subtractBackground(bg = stillBack1) # use 'old' background
partIden1 <- identifyParticles(sbg = allImages1,threshold = 0.2,
                              select = "light",
                              pixelRange = c(5,500))
summary(partIden1)$n # all larger than 0, and all frames up to first
observed 0

I haven't tested this in detail, but I think this should work.
You could also do a combination of the above, e.g. deleting a frame if
there is one zero, but when there are X 0's in a row, stop analyzing.

Good luck!

Marjolein

Op ma 7 jun. 2021 om 13:56 schreef haasdavid ***@***.***>:

**Comment by haasdavid** [comment]
Dear Marjolein,

thank you very much. This worked. And sorry for the long time without reply. I had to interrupt coding to acquire my data :-) 
But as soon as I startet agein... I ran into new problems...

Since my Images only have one color channel I get the following warning:

> Warning message:
> In loadImages(dirPictures = "ImageSequences/20210614_2-5_1_1000sp_ps_1",  :
>   Only 1 color channel detected (grey scale image).

This would be no problem, but the plotting of data stored in the records file won't work this way, I fear. Is there an existing workaround for this too? Else I would have to work with the part Iden file and ggplot. I receive this error when I try to plot the results:
[20210729_Anschreiben_David_Haas_Fraunhofer_Angehender_Projektleiter_Entscheidungsunterstützung_Klimawandel.pdf](https://github.com/marjoleinbruijning/trackdem/files/7000401/20210729_Anschreiben_David_Haas_Fraunhofer_Angehender_Projektleiter_Entscheidungsunterstutzung_Klimawandel.pdf)


> Error in .local(x, ...) : not a valid subset
> In addition: Warning message:
> In plot.TrDm(colorimages, frame = frame) :
>   Creating plot for 1 color channel image
